### PR TITLE
v0.90.0 update of opentelemetry-operator

### PR DESCRIPTION
**Pull request by lisguo**
*Description of changes:*
Pulls in all changes from v0.90.0 of the upstream opentelemetry-operator code. This is based off of of the upstream branch, with the amazon-cloudwatch-agent-operator/main branch merged on top. 

New changes:
- /pkg/collector was moved to /internal/manifests/collector
- add support for apachehttpd, dotnet, golang, nginx, nodejs, python (but would still need to be enabled via ACWA CR)
- add support for healthchecks (still needs some steps for enablement)
- add support for watching statefulsets, horizontalpodautoscaler,poddisruptionbudget
```
		Owns(&corev1.ConfigMap{}).
		Owns(&corev1.ServiceAccount{}).
		Owns(&corev1.Service{}).
		Owns(&appsv1.Deployment{}).
		Owns(&appsv1.DaemonSet{}).
		Owns(&appsv1.StatefulSet{}).
		Owns(&autoscalingv2.HorizontalPodAutoscaler{}).
		Owns(&policyV1.PodDisruptionBudget{})
```
FYI the above change requires changes to the clusterrole to list the new resources. This is reflected in `config/rbac` but needs to be added to the helm chart.
- all unit tests were migrated from upstream (still needs some work to run)

Renamed the following:
- opentelemetry-operator -> amazon-cloudwatch-agent-operator
- OpenTelemetryCollector -> AmazonCloudWatchAgent
- opentelemetry.io -> cloudwatch.aws.amazon.com

Things that were unchanged:
- versions.txt 
- ecr repo for the agent and adot instrumentation are still hardcoded in main.go
- default instrumentation logic in podmutator.go
- everything in /config, /helm, /integration-tests, /.github

As part of this upgrade, there was a [commit](https://github.com/open-telemetry/opentelemetry-operator/pull/1273/files) made to migrate the securityContext from the app container to the init container. This however caused issues when I tried to use the [sample app](https://github.com/aws-observability/application-signals-demo/blob/main/scripts/eks/appsignals/sample-app/api-gateway-deployment.yaml) in the appsignals-demo. Potentially related to this [issue](https://github.com/open-telemetry/opentelemetry-operator/issues/2272). To resolve this I commented out the following line in `pkg/instrumentation/sdk.go`

```
pod = i.setInitContainerSecurityContext(pod, pod.Spec.Containers[index].SecurityContext, javaInitContainerName)
```

*Testing*
I deployed the operator, crds, and clusterroles to my EKS cluster using `make deploy`. I deployed the sample app in the [application-signals-demo](https://github.com/aws-observability/application-signals-demo/blob/main/scripts/eks/appsignals/deploy-sample-app.sh). I verified that the java instrumentation sdk is injected:

```
Init Containers:
  opentelemetry-auto-instrumentation-java:
    Container ID:  containerd://31a8b52c6371918bb3af4db57623e5eadc1f5c8a301b193353b27b95313499f0
    Image:         public.ecr.aws/aws-observability/adot-autoinstrumentation-java:v1.31.1
    Image ID:      public.ecr.aws/aws-observability/adot-autoinstrumentation-java@sha256:08d05d413361ea2a11e27b4392f6fe856e384136e8d28f995cfc98438aa8a3f2
    Port:          <none>
    Host Port:     <none>
    Command:
      cp
      /javaagent.jar
      /otel-auto-instrumentation-java/javaagent.jar
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 09 Jan 2024 23:00:17 -0500
      Finished:     Tue, 09 Jan 2024 23:00:17 -0500
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /otel-auto-instrumentation-java from opentelemetry-auto-instrumentation-java (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-rkqx7 (ro)
```

I also applied the AmazonCloudWatchAgent resource and saw the agent running
```
amazon-cloudwatch   cloudwatch-agent-47dph                          1/1     Running   0              12m
amazon-cloudwatch   cloudwatch-agent-zv9pf                          1/1     Running   0              12m
```
logs:
```
2024-01-10T04:48:16Z I! {"caller":"otlpreceiver@v0.84.0/otlp.go:83","msg":"Starting GRPC server","kind":"receiver","name":"otlp/app_signals","data_type":"metrics","endpoint":"0.0.0.0:4315"}
2024-01-10T04:48:16Z W! {"caller":"internal@v0.84.1-0.20230908201109-ab3d6c5b6470/warning.go:40","msg":"Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks","kind":"receiver","name":"otlp/app_signals","data_type":"metrics","documentation":"https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks"}
2024-01-10T04:48:16Z I! {"caller":"otlpreceiver@v0.84.0/otlp.go:101","msg":"Starting HTTP server","kind":"receiver","name":"otlp/app_signals","data_type":"metrics","endpoint":"0.0.0.0:4316"}
2024-01-10T04:48:16Z I! {"caller":"service/service.go:161","msg":"Everything is ready. Begin running and processing data."}
```

services:
```
cloudwatch-agent              ClusterIP   10.100.22.172    <none>        4315/TCP,4316/TCP,2000/TCP   57m
cloudwatch-agent-headless     ClusterIP   None             <none>        4315/TCP,4316/TCP,2000/TCP   57m
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: config/manager/kustomization.yaml [hunk 1]
```diff
@@ -4,5 +4,5 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
 - name: controller
-  newName: 506463145083.dkr.ecr.us-west-2.amazonaws.com/cwagent-operator-pre-release
+  newName: 355660395157.dkr.ecr.us-east-1.amazonaws.com/amazon-cloudwatch-agent-operator
```
Comment (mitali-salvi) [inline config/manager/kustomization.yaml:7]: Is this intentional ? This seems like a personal account

File: config/prometheus/monitor.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+
+# Prometheus Monitor Service (Metrics)
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
```
Comment (mitali-salvi) [inline config/prometheus/monitor.yaml:4]: Just curious, will this set up prometheus monitoring for the operator ?

File: config/webhook/manifests.yaml [hunk 3]
```diff
@@ -15,7 +15,7 @@ webhooks:
   name: mamazoncloudwatchagent.kb.io
   rules:
   - apiGroups:
-    - cloudwatch.aws.amazon.com
+    - cloudwatch.aws.amazon.co
```
Comment (mitali-salvi) [inline config/webhook/manifests.yaml:18]: Typo ???

File: controllers/opentelemetrycollector_controller.go
Comment (mitali-salvi) [inline controllers/opentelemetrycollector_controller.go:1]: Is this rename intentional ? I think it should remain the amazoncloudwatchagent_controller.go

File: internal/manifests/collector/adapters/config_from.go [hunk 4]
```diff
@@ -5,16 +5,14 @@
...
 	"errors"
 
 	"gopkg.in/yaml.v2"
 )
 
 var (
 	// ErrInvalidYAML represents an error in the format of the configuration file.
-	ErrInvalidYAML = errors.New("couldn't parse the yaml configuration")
-	ErrInvalidJSON = errors.New("couldn't parse cloudwatch agent json configuration")
+	ErrInvalidYAML = errors.New("couldn't parse the opentelemetry-collector configuration")
```
Comment (mitali-salvi) [inline internal/manifests/collector/adapters/config_from.go:15]: Should this be the cloudwatch agent config ?

File: internal/manifests/collector/parser/receiver/receiver_aws-xray.go
Comment (mitali-salvi) [inline internal/manifests/collector/parser/receiver/receiver_aws-xray.go:1]: Do we need the receiver package ? I am guessing these files contain logic related to otel receivers

File: main.go [hunk 5]
```diff
@@ -10,29 +10,32 @@ import (
...
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
-	cwv1alphav1 "github.com/aws/amazon-cloudwatch-agent-operator/apis/v1alpha1"
+	otelv1alpha1 "github.com/aws/amazon-cloudwatch-agent-operator/apis/v1alpha1"
```
Comment (mitali-salvi) [inline main.go:30]: maybe nitpick - but this can be put back to its original value

File: main.go [hunk 6]
```diff
@@ -69,36 +83,42 @@ func main() {
...
 		tlsOpt                  tlsConfig
 	)
 
-	pflag.StringVar(&agentImage, "agent-image", fmt.Sprintf("%s:%s", cloudwatchAgentImageRepository, v.AmazonCloudWatchAgent), "The default cloudwatch agent image. This image is used when no image is specified in the CustomResource.")
-	pflag.StringVar(&autoInstrumentationJava, "auto-instrumentation-java-image", fmt.Sprintf("%s:%s", autoInstrumentationJavaImageRepository, v.AutoInstrumentationJava), "The default OpenTelemetry Java instrumentation image. This image is used when no image is specified in the CustomResource.")
+	pflag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
+	pflag.StringVar(&probeAddr, "health-probe-addr", ":8081", "The address the probe endpoint binds to.")
+	pflag.StringVar(&pprofAddr, "pprof-addr", "", "The address to expose the pprof server. Default is empty string which disables the pprof server.")
+	stringFlagOrEnv(&collectorImage, "collector-image", "RELATED_IMAGE_COLLECTOR", fmt.Sprintf("%s:%s", cloudwatchAgentImageRepository, v.AmazonCloudWatchAgent), "The default OpenTelemetry collector image. This image is used when no image is specified in the CustomResource.")
```
Comment (mitali-salvi) [inline main.go:98]: This should be corrected to "The default cloudwatch agent image. This image is used when no image is specified in the CustomResource."

File: Dockerfile [hunk 7]
```diff
@@ -27,7 +27,7 @@ ARG AGENT_VERSION
 ARG AUTO_INSTRUMENTATION_JAVA_VERSION
 
 # Build
-RUN CGO_ENABLED=0 GOOS=linux GO111MODULE=on go build -ldflags="-X ${VERSION_PKG}.version=${VERSION} -X ${VERSION_PKG}.buildDate=${VERSION_DATE} -X ${VERSION_PKG}.agent=${AGENT_VERSION} -X ${VERSION_PKG}.autoInstrumentationJava=${AUTO_INSTRUMENTATION_JAVA_VERSION}" -a -o manager main.go
+RUN CGO_ENABLED=0 GOOS=linux GO111MODULE=on go build -ldflags="-X ${VERSION_PKG}.version=${VERSION} -X ${VERSION_PKG}.buildDate=${VERSION_DATE} -X ${VERSION_PKG}.otelCol=${AGENT_VERSION} -X ${VERSION_PKG}.autoInstrumentationJava=${AUTO_INSTRUMENTATION_JAVA_VERSION}" -a -o manager main.go
```
Comment (mitali-salvi) [inline Dockerfile:30]: should this agent instead of otelCol ?

File: config/manager/kustomization.yaml (hunk 1 above)
Comment (lisguo) [reply config/manager/kustomization.yaml:7]: good catch, actually both of these images should be changed to use the public image since the previous image was also private

File: config/prometheus/monitor.yaml (hunk 2 above)
Comment (lisguo) [reply config/prometheus/monitor.yaml:4]: I can remove config/prometheus since we don't support it

File: controllers/opentelemetrycollector_controller.go
Comment (lisguo) [reply controllers/opentelemetrycollector_controller.go:1]: This rename is intentional for future otel upgrades. It's hard to view the diffs otherwise so I opted to keep the name of the files with upstream.

File: internal/manifests/collector/adapters/config_from.go (hunk 4 above)
Comment (lisguo) [reply internal/manifests/collector/adapters/config_from.go:15]: I think I missed the JSON validation logic here, good catch

File: Dockerfile (hunk 7 above)
Comment (lisguo) [reply Dockerfile:30]: I set it back to otelCol to by in sync with upstream. The variable is `otelCol` in main.go which is consistent.

File: Makefile [hunk 8]
```diff
@@ -115,7 +115,7 @@ set-image-controller: manifests kustomize
 # Deploy controller in the current Kubernetes context, configured in ~/.kube/config
 .PHONY: deploy
 deploy: set-image-controller
-	$(KUSTOMIZE) build config/default | kubectl apply -f -
+	$(KUSTOMIZE) build config/default | kubectl apply --server-side -f -
```
Comment (mitali-salvi) [inline Makefile:118]: the --server-side flag is such a cool feature

File: config/webhook/manifests.yaml [hunk 9]
```diff
@@ -81,7 +81,7 @@ webhooks:
   name: vamazoncloudwatchagentcreateupdate.kb.io
   rules:
   - apiGroups:
-    - cloudwatch.aws.amazon.com
+    - cloudwatch.aws.amazon.co
```
Comment (mitali-salvi) [inline config/webhook/manifests.yaml:84]: Typo again ?

File: config/webhook/manifests.yaml [hunk 10]
```diff
@@ -101,7 +101,7 @@ webhooks:
   name: vamazoncloudwatchagentdelete.kb.io
   rules:
   - apiGroups:
-    - cloudwatch.aws.amazon.com
+    - cloudwatch.aws.amazon.co
```
Comment (mitali-salvi) [inline config/webhook/manifests.yaml:104]: Here as well

File: config/crd/bases/cloudwatch.aws.amazon.com_amazoncloudwatchagents.yaml
Comment (mitali-salvi) [inline config/crd/bases/cloudwatch.aws.amazon.com_amazoncloudwatchagents.yaml:1]: Should this file be named the opentelemetrycollector.yaml to follow the naming convention in the PR ?

Comment (lisguo) [reply config/crd/bases/cloudwatch.aws.amazon.com_amazoncloudwatchagents.yaml:1]: The name AmazonCloudWatchAgent for the CRD will need to be different due to conflict with customers who will mistake our operator for the otel operator.

Any variables or artifacts that are specific to the code and are not customer facing are left as "otel" or "opentelemetry". However resources that can be seen by customers on their cluster should be "amazon cloudwatch agent" or "cloudwatch"/"agent".
